File: models/Gpt-3.5-turbo/cpp/code/problems/advanced_techniques/Necessary_Cities.cpp

#include <iostream>
#include <vector>

using namespace std;

vector<vector<int>> adj;
vector<bool> visited;

void dfs(int v) {
    visited[v] = true;
    for (int u : adj[v]) {
        if (!visited[u]) {
            dfs(u);
        }
    }
}

int main() {
    int n, m;
    cin >> n >> m;

    adj.resize(n + 1);
    visited.resize(n + 1, false);

    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;
        adj[a].push_back(b);
        adj[b].push_back(a);
    }

    for (int i = 1; i <= n; i++) {
        if (!visited[i]) {
            dfs(i);
        }
    }

    vector<int> necessaryCities;
    for (int i = 1; i <= n; i++) {
        if (!visited[i]) {
            necessaryCities.push_back(i);
        }
    }

    cout << necessaryCities.size() << endl;
    for (int city : necessaryCities) {
        cout << city << " ";
    }
    cout << endl;

    return 0;
}

File: models/Gpt-3.5-turbo/py/code/advanced_techniques/Necessary_Cities.py

def dfs(city):
    visited[city] = True
    for adj_city in adj_list[city]:
        if not visited[adj_city]:
            dfs(adj_city)

n, m = map(int, input().split())

adj_list = [[] for _ in range(n + 1)]
visited = [False] * (n + 1)

for _ in range(m):
    a, b = map(int, input().split())
    adj_list[a].append(b)
    adj_list[b].append(a)

necessary_cities = []

for city in range(1, n + 1):
    if not visited[city]:
        necessary_cities.append(city)
        dfs(city)

print(len(necessary_cities))
print(*necessary_cities)
